sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubMessageToRow.java

```diff
@@ -95,28 +109,40 @@ public void processElement(ProcessContext context) {
    * payload, and attributes.
    */
   private List<Object> getFieldValues(ProcessContext context) {
+    Row payload = parsePayloadJsonRow(context.element());
     return messageSchema().getFields().stream()
-        .map(field -> getValueForField(field, context.timestamp(), context.element()))
+        .map(
+            field ->
+                getValueForField(
+                    field, context.timestamp(), context.element().getAttributeMap(), payload))
         .collect(toList());
   }
 
   private Object getValueForField(
-      Schema.Field field, Instant timestamp, PubsubMessage pubsubMessage) {
-
-    switch (field.getName()) {
-      case TIMESTAMP_FIELD:
+      Schema.Field field, Instant timestamp, Map<String, String> attributeMap, Row payload) {
+    // TODO: do this check once at construction time, rather than for every element.
```

Comment:
I imagine you just fork the DoFn and share utility code?

Reply:
Well what I wanted to do was make `PubsubMessageToRow` extend `PTransform` and switch out the implementation when expanding, but I keep getting hung up on the mutli-output/dlq stuff.